Does not build the tests with WITH_STL=CXX17

Linux: debian bookworm, GCC 12.2.0

With

```
git clean -fdx build

cmake -S . -B build                    \
  -DBUILD_TESTING=ON                   \
  -DWITH_EXAMPLES=OFF                  \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo    \
  -DCMAKE_POSITION_INDEPENDENT_CODE=ON \
  -DBUILD_SHARED_LIBS=ON               \
  -DWITH_OTLP_GRPC=ON                  \
  -DWITH_OTLP_HTTP=ON                  \
  -DWITH_ABSEIL=ON                     \
  -DWITH_BENCHMARK=OFF                 \
  -DCMAKE_CXX_STANDARD=17              \
  -DWITH_STL=CXX17

cmake --build build --target install
```

I get

```[ 90%] Linking CXX executable otlp_http_exporter_test
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `typeinfo for testing::internal::UntypedFunctionMockerBase'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::Log(testing::internal::LogSeverity, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::UntypedFunctionMockerBase::~UntypedFunctionMockerBase()'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::UntypedFunctionMockerBase::SetOwnerAndName(void const*, char const*)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::ReportUninterestingCall(testing::internal::CallReaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::GetFailureReporter()'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::ExpectationBase::AllPrerequisitesAreSatisfied() const'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::ExpectationBase::FindUnsatisfiedPrerequisites(testing::ExpectationSet*) const'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::UntypedFunctionMockerBase::Name() const'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::Mock::UnregisterLocked(testing::internal::UntypedFunctionMockerBase*)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::ExpectationBase::CheckActionCountIfNotDone() const'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::IllegalDoDefault(char const*, int)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::Mock::GetReactionOnUninterestingCalls(void const*)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::ExpectationBase::RetireAllPreRequisites()'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::UntypedFunctionMockerBase::MockObject() const'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::ExpectationBase::DescribeCallCountTo(std::ostream*) const'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::UntypedFunctionMockerBase::VerifyAndClearExpectationsLocked()'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::LogIsVisible(testing::internal::LogSeverity)'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::g_gmock_mutex'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::internal::UntypedFunctionMockerBase::UntypedFunctionMockerBase()'
/usr/bin/ld: ../../test_common/src/http/client/nosend/libopentelemetry_http_client_nosend.so: undefined reference to `testing::Expectation::~Expectation()'
collect2: error: ld returned 1 exit status
gmake[2]: *** [exporters/otlp/CMakeFiles/otlp_http_exporter_test.dir/build.make:125: exporters/otlp/otlp_http_exporter_test] Fehler 1
gmake[1]: *** [CMakeFiles/Makefile2:5466: exporters/otlp/CMakeFiles/otlp_http_exporter_test.dir/all] Fehler 2
gmake: *** [Makefile:146: all] Fehler 2
```

--- a/ext/test/http/CMakeLists.txt
+++ b/ext/test/http/CMakeLists.txt
@@ -5,7 +5,7 @@ if(WITH_HTTP_CLIENT_CURL)
   set(FILENAME curl_http_test)
   add_compile_definitions(WITH_CURL)
   add_executable(${FILENAME} ${FILENAME}.cc)
-  target_link_libraries(${FILENAME} ${GTEST_BOTH_LIBRARIES}
+  target_link_libraries(${FILENAME} ${GMOCK_LIB} ${GTEST_BOTH_LIBRARIES}
                         ${CMAKE_THREAD_LIBS_INIT})
 
   if(TARGET CURL::libcurl)
@@ -24,8 +24,8 @@ endif()
 
 set(URL_PARSER_FILENAME url_parser_test)
 add_executable(${URL_PARSER_FILENAME} ${URL_PARSER_FILENAME}.cc)
-target_link_libraries(${URL_PARSER_FILENAME} ${GTEST_BOTH_LIBRARIES}
-                      ${CMAKE_THREAD_LIBS_INIT} opentelemetry_api)
+target_link_libraries(${URL_PARSER_FILENAME} opentelemetry_api ${GMOCK_LIB}
+                      ${GTEST_BOTH_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})
 gtest_add_tests(
   TARGET ${URL_PARSER_FILENAME}
   TEST_PREFIX ext.http.urlparser.

--- a/test_common/src/http/client/nosend/CMakeLists.txt
+++ b/test_common/src/http/client/nosend/CMakeLists.txt
@@ -27,8 +27,7 @@ if(${BUILD_TESTING})
     find_library(GMOCK_LIB gmock PATH_SUFFIXES lib)
   endif()
 
-  target_link_libraries(
-    opentelemetry_http_client_nosend ${GTEST_BOTH_LIBRARIES} opentelemetry_ext
-    opentelemetry_test_common)
+  target_link_libraries(opentelemetry_http_client_nosend opentelemetry_ext
+                        opentelemetry_test_common ${GTEST_BOTH_LIBRARIES})
 
 endif()

--- a/test_common/src/http/client/nosend/CMakeLists.txt
+++ b/test_common/src/http/client/nosend/CMakeLists.txt
@@ -27,7 +27,8 @@ if(${BUILD_TESTING})
     find_library(GMOCK_LIB gmock PATH_SUFFIXES lib)
   endif()
 
-  target_link_libraries(opentelemetry_http_client_nosend opentelemetry_ext
-                        opentelemetry_test_common ${GTEST_BOTH_LIBRARIES})
+  target_link_libraries(
+    opentelemetry_http_client_nosend opentelemetry_ext
+    opentelemetry_test_common ${GMOCK_LIB} ${GTEST_BOTH_LIBRARIES})
 
 endif()

